File: library/src/main/java/com/cardinalblue/gesture/MyMotionEvent.java
//  Copyright Oct 2017-present CardinalBlue
//
//  Author: boy@cardinalblue.com
//          jack.huang@cardinalblue.com
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package com.cardinalblue.gesture;

public class MyMotionEvent {

    public final int maskedAction;

    public final int downPointerCount;
    public final float[] downXs;
    public final float[] downYs;

    public final boolean isUp;
    public final float upX;
    public final float upY;

    public MyMotionEvent(int maskedAction,
                         float[] downXs,
                         float[] downYs,
                         boolean isUp,
                         float upX,
                         float upY) {
        if (downXs == null || downYs == null) {
            throw new IllegalArgumentException("Invalid down x and y array.");
        } else if (downXs.length != downYs.length) {
            throw new IllegalArgumentException("Amount of down x is not consistent to y.");
        }

        this.maskedAction = maskedAction;

        this.downXs = downXs;
        this.downYs = downYs;
        this.downPointerCount = downXs.length;

        this.isUp = isUp;
        this.upX = upX;
        this.upY = upY;
    }
}

File: collage-gesture-detector/src/main/java/com/cardinalblue/gesture/MyMotionEvent.kt
// Copyright Feb 2017-present CardinalBlue
//
// Author: boy@cardinalblue.com
//         jack.huang@cardinalblue.com
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

package com.cardinalblue.gesture

class MyMotionEvent(val maskedAction: Int,
                    val downXs: FloatArray?,
                    val downYs: FloatArray?,
                    val isUp: Boolean,
                    val upX: Float,
                    val upY: Float) {

    val downPointerCount: Int
    val downFocusX: Float
    val downFocusY: Float

    init {
        if (downXs == null || downYs == null) {
            throw IllegalArgumentException("Invalid down x and y array.")
        } else if (downXs.size != downYs.size) {
            throw IllegalArgumentException("Amount of down x is not consistent to y.")
        }
        this.downPointerCount = downXs.size

        var sumX = 0f
        var sumY = 0f
        for (i in 0 until this.downPointerCount) {
            sumX += downXs[i]
            sumY += downYs[i]
        }
        this.downFocusX = sumX / this.downPointerCount
        this.downFocusY = sumY / this.downPointerCount
    }
}
